### PR TITLE
New data and usage of hashes

### PR DESCRIPTION
**Pull request by schmijos**

### CODE REVIEW

File: lib/swiss_village_directory/repository.rb [hunk 1]
```diff
@@ -6,9 +6,27 @@ class Repository
     attr_reader :villages
 
     def initialize
-      raw_data = CSV.read(File.dirname(__FILE__) + '/../../data/PLZO_CSV_WGS84.csv', col_sep: ';')
-      @villages = raw_data[1..-1].map do |row|
-        Village.new(row[0], row[1], row[2], row[3], row[4], row[5], row[6])
+      filename = File.dirname(__FILE__) + '/../../data/PLZO_CSV_WGS84.csv'
+      options = { col_sep: ';', encoding: 'ISO-8859-1:UTF-8', headers: :first_row }
```
Comment (schmijos) [inline lib/swiss_village_directory/repository.rb:10]: Last time I converted the source file before using it. Now I take it as-is.

File: spec/swiss_village_directory/repository_spec.rb [hunk 2]
```diff
@@ -1,10 +1,12 @@
 require 'spec_helper'
 
+cached_repository = SwissVillageDirectory::Repository.new
```
Comment (schmijos) [inline spec/swiss_village_directory/repository_spec.rb:3]: Like this the file is only read once.

File: spec/swiss_village_directory/repository_spec.rb [hunk 3]
```diff
@@ -13,53 +15,42 @@
...
+    expect(village.latitude).to eq(47.491)
   end
 
   describe '#find_all_by' do
-    before do
-      fake_csv = [['first_ignored_row_with_headers'],
-                  ['Aadorf', '8355', '0', 'AadorfCommune', 'TG', '8.906', '47.496'],
-                  ['Bbdorf', '8354', '0', 'AadorfCommune', 'AG', '9.906', '48.496']]
-      expect_any_instance_of(CSV).to receive(:read).and_return(fake_csv)
-    end
```
Comment (schmijos) [inline spec/swiss_village_directory/repository_spec.rb:36]: I go for the real data now.

File: lib/swiss_village_directory/repository.rb [hunk 4]
```diff
@@ -6,9 +6,27 @@ class Repository
...
+        'N' => :latitude,
+        'Sprache' => :language
+      }
+
+      villages_hashes = CSV.open(filename, options).map do |row|
+        row.to_h.transform_keys {|k| source_key_mapping[k] }
+      end
+
+      @villages = villages_hashes.map do |h|
+        Village.new(h[:name], h[:zip_code], h[:one_digit_spare], h[:commune], h[:canton], h[:longitude], h[:latitude])
```
Comment (schmijos) [inline lib/swiss_village_directory/repository.rb:29]: Language and BFS-Nr are not yet published to the interface. We should maybe reorganize this anyways.

File: lib/swiss_village_directory/repository.rb [hunk 5]
```diff
@@ -6,9 +6,27 @@ class Repository
     attr_reader :villages
 
     def initialize
-      raw_data = CSV.read(File.dirname(__FILE__) + '/../../data/PLZO_CSV_WGS84.csv', col_sep: ';')
-      @villages = raw_data[1..-1].map do |row|
-        Village.new(row[0], row[1], row[2], row[3], row[4], row[5], row[6])
+      filename = File.dirname(__FILE__) + '/../../data/PLZO_CSV_WGS84.csv'
+      options = { col_sep: ';', encoding: 'ISO-8859-1:UTF-8', headers: :first_row }
+
+      source_key_mapping = {
```
Comment (sihu) [inline lib/swiss_village_directory/repository.rb:12]: could be extracted to a separate function

File: lib/swiss_village_directory/repository.rb [hunk 6]
```diff
@@ -6,9 +6,27 @@ class Repository
...
+        'Zusatzziffer' => :one_digit_spare,
+        'Gemeindename' => :commune,
+        'BFS-Nr' => :bfs_number,
+        'Kantonskürzel' => :canton,
+        'E' => :longitude,
+        'N' => :latitude,
+        'Sprache' => :language
+      }
+
+      villages_hashes = CSV.open(filename, options).map do |row|
```
Comment (sihu) [inline lib/swiss_village_directory/repository.rb:24]: also this could be a function

File: spec/swiss_village_directory/repository_spec.rb (hunk 2 above)
Comment (sihu) [reply spec/swiss_village_directory/repository_spec.rb:3]: 👍 totally agree, makes sense

**Review by sihu — APPROVED** [review]
Really nice 👍 Love it ❤️

**Pull request merged**